### DISCUSSION TITLE
Workers ain't ... working for me

### DISCUSSION THREAD

**Discussion by valerio-afk** [Textualize/textual#3401] [General]
Hello,

I have very recently got to know Textual and of course I am giving a try by making a wee project (well, it's getting bigger than I thought, but just because I was unaware of the potential of this API).

I show you a code snippet below :

```
    @on(Button.Pressed,"#work_launcher")
    async def work_launcher_pressed(this,event):
        if this.is_analysing:
            this._update_job_related_interface()
            this.set_status("[bright_magenta]Operation stopped[/]")
            for w in this.workers:
                if w.name == "comparison":
                    w.cancel()

        else:
          # bla bla bla do stuff here

            match this.syncmode:
                case SyncMode.UPDATE:
                    this.run_worker(this._run_update,name="comparison", exclusive=True)


                case _:
                    raise NotImplementedError("Sync mode not implemented yet!")


    #@work(exclusive=True,name="comparison")
    async def _run_update(this):
        res = compare_tree(this.src, this.dst, this._backend)
```

With this configuration, the compare_tree function (which is very looong), runs synchronously. The interface freezes but I hear my CPU fan spinning. I tried with and without the decorator (when using the decorator, I also get rid of the this.run_worker). Nothing, it doesn't work. The only way I got it to work is to set threaded=True. Although it works, I cannot make it stop, meaning that the w.cancel() doesn't work. I am sure that the code gets into that call, because I debugged it a little bit.

Long story short:

if threaded=False, doesn't run asynchronously.
if threaded=True, does run asynchronously. but cannot get it cancelled.

Any ideas? I am nearly sure that I am the problem here. However, I am following the tutorial, the examples, but I cannot see where my mistake is.

**Comment by TomJGooding** [comment]
I assume your `compare_tree` function is not asynchronous (i.e. a regular function)? [That's why you may need to use threads](https://textual.textualize.io/guide/workers/#thread-workers).

**Reply by davep** [reply]
And further to what Tom says, regarding:

> Although it works, I cannot make it stop

Have you coded it so that it will stop? If not, see the [threaded worker example](https://textual.textualize.io/guide/workers/#thread-workers) and look at the use of the `is_cancelled` property.

**Comment by valerio-afk** [comment]
Hi both,

Thanks a lot for your suggestions and link to resources.
Checking the cancelled property in threaded worker is making its effect.

Cheers!
